Awd

cgraywang opened the pull request:
## Description ##
The awd-lm model implementation is almost done. But I encounter the following error msg, I have spent sometime on resolving this but still encounter the msg, in order to make it fixed faster, I would like @szha to help to take a look. Thanks.

Traceback (most recent call last):
  File "train.py", line 122, in <module>
    args.tied)
  File "/home/ubuntu/cgraywang/incubator-mxnet/python/mxnet/gluon/model_zoo/text/lm.py", line 37, in __init__
    super(AWDLSTM, self).__init__(vocab, vocab, **kwargs)
  File "/home/ubuntu/cgraywang/incubator-mxnet/python/mxnet/gluon/model_zoo/text/base.py", line 30, in __init__
    self.encoder = self._get_encoder()
  File "/home/ubuntu/cgraywang/incubator-mxnet/python/mxnet/gluon/model_zoo/text/lm.py", line 57, in _get_encoder
    else self._embed_dim, 0, self._weight_drop))
  File "/home/ubuntu/cgraywang/incubator-mxnet/python/mxnet/gluon/model_zoo/text/base.py", line 129, in get_rnn_layer
    _apply_weight_drop_to_rnn_layer(block, rate = weight_dropout, mode = 'training')
  File "/home/ubuntu/cgraywang/incubator-mxnet/python/mxnet/gluon/model_zoo/text/base.py", line 98, in _apply_weight_drop_to_rnn_layer
    block.params[key] = weight_dropped_params.data()
  File "/home/ubuntu/cgraywang/incubator-mxnet/python/mxnet/gluon/model_zoo/text/base.py", line 217, in data
    d = self._check_and_get(self._data, ctx)
  File "/home/ubuntu/cgraywang/incubator-mxnet/python/mxnet/gluon/parameter.py", line 189, in _check_and_get
    "nested child Blocks"%(self.name))
RuntimeError: Parameter awdlstm0_extendedsequential0_lstm0_l0_h2h_weight has not been initialized. Note that you should initialize parameters and create Trainer with Block.collect_params() instead of Block.params because the later does not include Parameters of nested child Blocks

szha commented on example/gluon/word_language_model/train.py:
```diff
@@ -22,11 +22,13 @@
 from mxnet.gluon import contrib
 from mxnet.gluon.model_zoo.text.lm import RNNModel
-import data
+from mxnet.gluon.model_zoo.text.lm import AWDLSTM
```
from XXX import RNNModel, AWDLSTM

szha commented on example/gluon/word_language_model/train.py:
```diff
@@ -22,11 +22,13 @@
 parser.add_argument('--model', type=str, default='lstm',
                     help='type of recurrent net (rnn_tanh, rnn_relu, lstm, gru)')
+parser.add_argument('--awd', type=str, default='true',
+                    help='Whether to apply awd (true, false)')
```
remove `--awd` flag, and add option for AWD model in the `--model` string instead.

cgraywang commented:
hi @szha , I have fixed the bug and addressed your comments. I have also started running the awd-lstm language model following the paper's suggested params. Hopefully, we can have some competitive results on Wiki-2 even some pieces are missing (optimization, embedding dropout, etc). Please have a look, and if it looks good to you. Please feel free to merge it into the master of origin. Thanks.

szha commented on python/mxnet/gluon/model_zoo/text/base.py:
```diff
@@ -74,22 +81,50 @@ def get_rnn_cell(mode, num_layers, num_hidden,
+        block = rnn.LSTM(num_hidden, num_layers, dropout=dropout,
+                        input_size=num_embed)
+        if weight_dropout:
+            _apply_weight_drop_to_rnn_layer(block, weight_dropout, weight_dropout_mode = weight_dropout_mode)
```
This function is a utility function for constructing RNN layers with a set of tricks. It doesn't need to know whether the block will be in awd or not. In that sense, you may move the weight_drop block outside of the last if condition and remove the awd-specific mode.

cgraywang replied:
The set of tricks is only for lstm, so I restrict the awd option only for lstm. We may want to hide the details to the model users, because we also do not the effectiveness of the model on other RNNs. It might cause unnecessary confusion for the users. How do you think?

szha commented on example/gluon/word_language_model/train.py:
```diff
@@ -106,10 +113,33 @@
+# type(model.params._params)
+# print("model.params._shared")
+# model.params._shared.context
+# type(model.params._shared)
```
clean up

szha commented on python/mxnet/gluon/model_zoo/text/base.py:
```diff
@@ -74,22 +81,50 @@ def get_rnn_cell(mode, num_layers, num_hidden,
+        if training:
+            _apply_weight_drop_to_rnn_cell(rnn_cell, rate = weight_dropout, mode = 'training')
+        else:
+            _apply_weight_drop_to_rnn_cell(rnn_cell, rate = weight_dropout, mode = 'always')
```
mode = weight_drop_mode

szha commented on python/mxnet/gluon/model_zoo/text/base.py:
```diff
@@ -74,22 +81,50 @@ def get_rnn_cell(mode, num_layers, num_hidden,
-        return rnn.RNN(num_hidden, 'relu', num_layers, dropout=dropout,
-                       input_size=num_embed)
+        block = rnn.RNN(num_hidden, 'relu', num_layers, dropout=dropout,
+                       input_size=num_embed)      
```
remove bleeding (whitespaces at the end)

szha commented on python/mxnet/gluon/model_zoo/text/base.py:
```diff
@@ -72,6 +72,8 @@ def get_rnn_cell(mode, num_layers, num_hidden,
             elif mode == 'gru':
                 cell = rnn.GRUCell(num_hidden)
+            elif mode == 'awd-lstm':
+                cell = rnn.LSTMCell(num_hidden)
```
New mode is not needed.

cgraywang replied:
you mean: elif mode == 'lstm' or mode == 'awd-lstm'?

szha replied:
there shouldn't be a mode specific for awd in this utility function. whether to apply weight drop or not should only depend whether the caller of this function requested non-zero weight-drop.

szha commented on example/gluon/word_language_model/train-weightdrop.py:
```diff
@@ -74,23 +74,23 @@
+print("--weight_dropout=" + str(args.weight_dropout))
+print("--tied=" + str(args.tied))
+print("--cuda=" + str(args.cuda))
+print("--eval_only=" + str(args.eval_only))
```
you can simply log `args`

szha commented on example/gluon/word_language_model/train.py:
```diff
@@ -159,7 +159,7 @@ def eval(data_source):
         for X, y, h in zip(data_list, target_list, hiddens):
             output, h = model(X, h)
-            Ls.append(loss(mx.nd.reshape(output, (-3, -1)), mx.nd.reshape(y, (-1,))))
+            Ls.append(loss(output.reshape((-3, -1)), y.reshape((-1,))))
```
array.reshape doesn't support extended flags such as -3 yet.
